File: Parallel_IO.cpp
#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>
#include <mutex>
#include <vector>
#include <numeric> // std::partial_sum

/*
This C++ program performs an I/O-bound benchmark by writing and reading a large number of integers to and from a file using parallel processing.
Due to the inherent sequential nature of file I/O, true parallelization is not possible,
but we can simulate some degree of concurrency by using separate threads for writing and reading.
*/

const int N = 5000000;
const std::string filename = "io_test_c_plus_plus_parallel.txt";

// Function to write numbers to a file (Thread 1)
void writeNumbersToFile() {
    std::ofstream outFile(filename);
    if (!outFile.is_open()) {
        std::cerr << "Error opening output file." << std::endl;
        return;
    }
    for (int i = 0; i < N; ++i) {
        outFile << i << "\n";
    }
    outFile.close();
}

// Function to read numbers from a file and calculate the sum (Thread 2)
long readNumbersFromFile() {
    std::ifstream inFile(filename);
    if (!inFile.is_open()) {
        std::cerr << "Error opening input file." << std::endl;
        return -1; // Indicate error
    }
    long sum = 0;
    int value;
    while (inFile >> value) {
        sum += value;
    }
    inFile.close();
    return sum;
}

int main() {
    // Start timing
    auto start = std::chrono::high_resolution_clock::now();
    std::clock_t c_start = std::clock();

    // Launch the threads for writing and reading
    std::thread writeThread(writeNumbersToFile);
    std::thread readThread(readNumbersFromFile);

    // Wait for the threads to complete
    writeThread.join();
    long sum = 0;
    //It only measures correctly if it waits until after the read thread has completed
    readThread.join();
    sum = readNumbersFromFile();

    // Print the sum (only if reading was successful)
    if (sum != -1) {
        std::cout << "Sum: " << sum << std::endl;
    } else {
        std::cerr << "Error calculating sum." << std::endl;
    }

    // End timing and calculate execution time
    std::clock_t c_end = std::clock();
    auto end = std::chrono::high_resolution_clock::now();

    // Calculate CPU and Execution time in milliseconds
    double cpu_time = 1000.0 * (c_end - c_start) / CLOCKS_PER_SEC;
    std::chrono::duration<double, std::milli> exec_time = end - start;

    std::cout << "CPU time: " << cpu_time << " ms" << std::endl;
    std::cout << "Execution time: " << exec_time.count() << " ms" << std::endl;

    return 0;
}

File: Parallel_IO.py
import time
import threading

'''
This Python program performs an I/O-bound benchmark by writing and reading a large number of integers to and from a file using threading.
Due to the GIL and inherent sequential nature of file I/O, true parallelization is not possible.
'''

def write_numbers_to_file(filename, N):
    with open(filename, "w") as f:
        for i in range(N):
            f.write(f"{i}\n")

def read_numbers_from_file(filename):
    total = 0
    try:
        with open(filename, "r") as f:
            for line in f:
                total += int(line.strip())
    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
        return None  # Indicate error
    return total

def main():
    # Start timing
    start_time = time.time()
    start_cpu_time = time.process_time()

    N = 5000000
    filename = "io_test_python_parallel.txt"

    # Launch the threads
    write_thread = threading.Thread(target=write_numbers_to_file, args=(filename, N))
    read_thread = threading.Thread(target=read_numbers_from_file, args=(filename,))

    write_thread.start()
    read_thread.start() #The IO can only be tested one at a time, so don't await the read function

    write_thread.join()
    #Important: the read thread must only be read after the write thread has completed writing.
    read_thread.join()

    #Check for errors, since this multithreaded function is highly prone to errors.
    result = read_numbers_from_file(filename)

    # End timing and calculate times
    end_cpu_time = time.process_time()
    end_time = time.time()

    # Convert times to milliseconds
    cpu_time = (end_cpu_time - start_cpu_time) * 1000
    exec_time = (end_time - start_time) * 1000

    # Print metrics
    print(f"CPU Time: {cpu_time:.2f} ms")
    print(f"Execution Time: {exec_time:.2f} ms")

    # Print the sum
    if result is not None:
        print(f"Sum: {result}")
    else:
        print("Error: Could not calculate the sum. May have failed to write to file.")
if __name__ == "__main__":
    main()